### PR TITLE
space-age: Add generator

### PR DESCRIPTION
**Pull request by ajwann**
## Why?

space-age exercise needs a generator

## What?

added generator for space-age

## How Has This Been Tested?

rake test:space-age
rubocop -D

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or enhancement that would cause existing functionality to change)
- [ ] Gardening (code and/or documentation cleanup)

## References and Closures

references #396

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change relies on a pending issue/pull request
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: exercises/space-age/space_age_test.rb [hunk 1]
```diff
@@ -1,30 +1,24 @@
...
 require_relative 'space_age'
 
+# Test data version: 7c63e40
 class SpaceAgeTest < Minitest::Test
   DELTA = 0.01
 
-  def test_age_in_seconds
-    age = SpaceAge.new(1_000_000)
-    assert_in_delta 1_000_000, age.seconds, DELTA
-  end
```
Comment (ajwann) [inline exercises/space-age/space_age_test.rb]: This method was missing from x-common, so I omitted it. If we need it, I can add it to x-common or manually hack it in before the ```test_age_on_earth``` method in space_age_cases.rb

**Comment by tommyschaefer** [comment]
Hi There! First off, thank you for your help converting this exercise to use generated tests!

There has been some really awesome new work on generators by @hilary that has changed how we implement new generators. Check out the [updated documentation](https://github.com/exercism/xruby#generated-test-suites), for some tips on upgrading this to the new pattern!

Thank you again for your work and time! 😄

**Comment by hilary** [comment]
You might also find it helpful to see what the final version of the etl PR looked like: exercism/xruby#575.

**Comment by ajwann** [comment]
Awesome, thank you both. I've read over the final etl PR and I'll take a look at the new docs as soon as I can this week.

File: exercises/space-age/space_age_test.rb (hunk 1 above)
Comment (Insti) [inline exercises/space-age/space_age_test.rb]: Leaving it out is fine. If it's not specified in the readme or the canonical data it probably shouldn't be there.

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+class SpaceAgeCase < ExerciseCase
+  SECONDS_PER_YEAR_ON_EARTH = 31557600
+
+  def workload
+    [
+      "age = SpaceAge.new(#{underscore_format(seconds)})",
+      indent(4, assertion),
+    ].join("\n")
+  end
```
Comment (ajwann) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: Now workload is the only public method as per the [docs](https://github.com/exercism/xruby#generated-test-suites).

File: exercises/space-age/.meta/generator/test_template.erb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env ruby
+gem 'minitest', '>= 5.0.0'
+require 'minitest/autorun'
+require_relative '<%= exercise_name %>'
+
+# Common test data version: <%= abbreviated_commit_hash %>
+class <%= exercise_name_camel %>Test < Minitest::Test
+  DELTA = 0.01
```
Comment (ajwann) [inline exercises/space-age/.meta/generator/test_template.erb]: I still needed a custom template to add this constant.

**Comment by ajwann** [comment]
I think I've moved everything where it belongs, please let me know if I missed anything.

File: exercises/space-age/space_age_test.rb [hunk 4]
```diff
@@ -3,28 +3,24 @@
...
+    # skip
     age = SpaceAge.new(1_000_000_000)
     assert_in_delta 31.69, age.on_earth, DELTA
   end
 
-  def test_age_in_mercury_years
+  def test_age_on_mercury
     skip
     age = SpaceAge.new(2_134_835_688)
     assert_in_delta 67.65, age.on_earth, DELTA
```
Comment (Insti) [inline exercises/space-age/space_age_test.rb]: I don't like that the `on_earth` method is also being tested here, I wouldn't do it in real life code, but I cannot tell if this test has educational value and should remain. 

https://github.com/exercism/xruby#workload-philosophy

Thoughts?

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
+class SpaceAgeCase < ExerciseCase
+  SECONDS_PER_YEAR_ON_EARTH = 31557600
+
+  def workload
+    [
+      "age = SpaceAge.new(#{underscore_format(seconds)})",
+      indent(4, assertion),
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: There are some (currently undocumented) methods in `ExerciseCases` that can do underscore formatting and indenting for you.

https://github.com/exercism/xruby/blob/master/lib/generator/exercise_cases.rb

https://github.com/exercism/xruby/blob/master/lib/generator/underscore.rb

File: exercises/space-age/.meta/generator/test_template.erb (hunk 3 above)
Comment (Insti) [inline exercises/space-age/.meta/generator/test_template.erb]: I've created an issue for this: https://github.com/exercism/xruby/issues/589

**Comment by Insti** [comment]
Sorry for all the complications here @ajwann you're on the bleeding edge of test generation right now! :hocho:

File: exercises/space-age/.meta/generator/test_template.erb (hunk 3 above)
Comment (Insti) [inline exercises/space-age/.meta/generator/test_template.erb]: I wonder if a comment explaining what delta is used for would be helpful for people who are meeting `assert_in_delta` for the first time.

Comment (ajwann) [inline exercises/space-age/.meta/generator/test_template.erb]: I think that would be helpful, I'll add it.

File: exercises/space-age/space_age_test.rb (hunk 4 above)
Comment (ajwann) [inline exercises/space-age/space_age_test.rb]: I don't think it provides a lot value, other than to give the user a baseline for which to compare the age difference. I personally think we could take out the on_earth assertions and just leave the ```age_on_earth``` test.

File: exercises/space-age/.meta/generator/space_age_cases.rb (hunk 5 above)
Comment (ajwann) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: Sounds good; I can use the ```indent_lines``` method of ```ExerciseCases```, and add the integer underscore formatting method I'm using to the ```Underscore``` module.

File: exercises/space-age/.meta/generator/test_template.erb (hunk 3 above)
Comment (ajwann) [inline exercises/space-age/.meta/generator/test_template.erb]: I've though about what this comment should say, and I'm kind of embarrassed to say that I don't really know what delta means in this case. Besides the basic mathematics definition of "change in a given quantity", I'm kind of stumped.

**Comment by ajwann** [comment]
I think we are finally good to go!

File: exercises/space-age/.meta/generator/test_template.erb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env ruby
+gem 'minitest', '>= 5.0.0'
```
Comment (Insti) [inline exercises/space-age/.meta/generator/test_template.erb]: This line is no longer used.

File: lib/generator/underscore.rb [hunk 7]
```diff
@@ -5,5 +5,11 @@ def underscore
         downcase.gsub(/[- ]/, '_').gsub(/[^\w?]/, '')
       end
     end
+
+    refine Fixnum do
+      def underscore
+        self.to_s.reverse.gsub(/...(?=.)/, '\&_').reverse
```
Comment (Insti) [inline lib/generator/underscore.rb]: This is good 👍

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+class SpaceAgeCase < Generator::ExerciseCase
+  using Generator::Underscore
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: Is `Generator::Underscore` is already used by `Generator::ExerciseCase`

**Comment by Insti** [comment]
This is looking good @ajwann , sorry we keep moving the goalposts on you!

File: exercises/space-age/space_age_test.rb [hunk 9]
```diff
@@ -1,67 +1,82 @@
 #!/usr/bin/env ruby
+gem 'minitest', '>= 5.0.0'
 require 'minitest/autorun'
 require_relative 'space_age'
 
+# Common test data version: 7c63e40
```
Comment (Insti) [inline exercises/space-age/space_age_test.rb]: This line is missing the `canonical_data_version`

Comment (Insti) [inline exercises/space-age/space_age_test.rb]: See: https://github.com/exercism/xruby/blob/master/lib/generator/test_template.erb

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
+class SpaceAgeCase < Generator::ExerciseCase
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: There should be a `require 'generator/exercise_case'` at the top of this file.

See: https://github.com/exercism/xruby#implementing-a-generator

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+      assertion,
+    ].join("\n")
+  end
+
+  private
+
+  def assertion
+    [
+      "    #{assertion_for_location(expected, planet)}",
+    ]
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: Does this need to be an array?

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def assertion_for_location(expected, location)
+    "assert_in_delta #{expected}, age.on_#{location.downcase}, DELTA"
+  end
+
+end
+
+SpaceAgeCases = proc do |data|
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: This section is no longer needed.

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  using Generator::Underscore
+
+  SECONDS_PER_YEAR_ON_EARTH = 31557600
+
+  def workload
+    [
+      "age = SpaceAge.new(#{seconds.underscore})",
+      assertion,
+    ].join("\n")
+  end
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: There might be a method in Generator::Exercise case that can join and indent for you.

File: exercises/space-age/space_age_test.rb (hunk 9 above)
Comment (Insti) [inline exercises/space-age/space_age_test.rb]: These comments should be about the `test_template.erb` version. (sorry)

File: exercises/space-age/space_age_test.rb [hunk 14]
```diff
@@ -1,67 +1,82 @@
...
 class SpaceAgeTest < Minitest::Test
+  # assert_in_delta will pass if the difference
+  # between the values being compared is less
+  # than the allowed delta
```
Comment (Insti) [inline exercises/space-age/space_age_test.rb]: Great explanatory comment. 👍

File: exercises/space-age/.meta/generator/space_age_cases.rb (hunk 8 above)
Comment (ajwann) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: Yep, it definitely is. I can swap my ```using``` statement in favor of requiring 'generator/exercise_case'```.

File: exercises/space-age/.meta/generator/space_age_cases.rb (hunk 13 above)
Comment (ajwann) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: I ended uup using ```indent_lines```.

Comment (ajwann) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: I ended up using ```indent_lines```.

**Comment by ajwann** [comment]
@Insti I've made all the requested updates, but I am having trouble testing the changes. I rebased off of master, but when I attempt to run ```bin/generate space-age``` I get an error saying that a generator currently doesn't exist for spage-age. My generator is still in ```exercises/space-age/.meta/generator/``` so I'm not sure what's causing the issue.

**Comment by Insti** [comment]
@ajwann I'll have a look into this for you when I get a chance.

File: exercises/space-age/space_age_test.rb [hunk 15]
```diff
@@ -1,67 +1,82 @@
 #!/usr/bin/env ruby
+gem 'minitest', '>= 5.0.0'
```
Comment (Insti) [inline exercises/space-age/space_age_test.rb]: This line is no longer required/used and can be deleted.

**Comment by Insti** [comment]
@ajwann 

> when I attempt to run bin/generate space-age I get an error saying that a generator currently doesn't exist for spage-age. My generator is still in exercises/space-age/.meta/generator/ so I'm not sure what's causing the issue.

The expected filename has changed, it should now be called `space_age_case.rb` (no 's' on 'case'). More apologies for changing everything on you!

File: exercises/space-age/.meta/generator/space_age_cases.rb (hunk 8 above)
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: It turns out you can not, due to lexical scope issues. Sorry for misleading you.

http://blog.honeybadger.io/understanding-ruby-refinements-and-lexical-scope/

File: exercises/space-age/.meta/generator/space_age_cases.rb [hunk 16]
```diff
@@ -0,0 +1,13 @@
+require 'generator/exercise_case'
+
+class SpaceAgeCase < Generator::ExerciseCase
+
+  SECONDS_PER_YEAR_ON_EARTH = 31557600
+
+  def workload
+    indent_lines(["age = SpaceAge.new(#{seconds.underscore})",
+      "    assert_in_delta #{expected}, age.on_#{planet.downcase}, DELTA"
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_cases.rb]: The spaces between `"` and `assert` are not required when you're using `indent_lines`.

File: exercises/space-age/.meta/generator/space_age_case.rb [hunk 17]
```diff
@@ -0,0 +1,12 @@
+require 'generator/exercise_case'
+
+class SpaceAgeCase < Generator::ExerciseCase
+  using Generator::Underscore
```
Comment (ajwann) [inline exercises/space-age/.meta/generator/space_age_case.rb]: I'm not too familiar with ```refinements```, but it looks like this using statement is necessary. 
It doesn't look like the using statement in ```ExerciseCase``` gets inherited by ```SpaceAgeCase```.

Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_case.rb]: Yeah, I also discovered this and commented here: https://github.com/exercism/xruby/pull/583#discussion_r116457149
> It turns out you can not, due to lexical scope issues. Sorry for misleading you.
> http://blog.honeybadger.io/understanding-ruby-refinements-and-lexical-scope/

Looks like github managed to hide the comment.

I also made an issue about addressing this here in xruby: https://github.com/exercism/xruby/issues/645

But including it for now is good. 👍

File: exercises/space-age/.meta/generator/space_age_case.rb [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+  SECONDS_PER_YEAR_ON_EARTH = 31557600
```
Comment (Insti) [inline exercises/space-age/.meta/generator/space_age_case.rb]: Is this constant used/necessary?

File: exercises/space-age/.meta/generator/space_age_case.rb [hunk 19]
```diff
@@ -0,0 +1,11 @@
+require 'generator/exercise_case'
+
+class SpaceAgeCase < Generator::ExerciseCase
+  using Generator::Underscore
+
```
Comment (ajwann) [inline exercises/space-age/.meta/generator/space_age_case.rb]: @Insti; I definitely didn't need that constant, it was left over from when each assertion was unnecessarily testing the age on Earth.

**Comment by Insti** [comment]
Awesome! Thanks for all your work on this @ajwann. I've merged it now, so any more changes we make to the process will be our responsibility to fix! 😄

**Pull request merged**